src/Workspaces/Core/Portable/SourceGeneratorTelemetry/SourceGeneratorTelemetryCollectorWorkspaceService.cs

```diff
@@ -0,0 +1,77 @@
+using System.Text;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.SourceGeneratorTelemetry
+{
+    internal class SourceGeneratorTelemetryCollectorWorkspaceService : ISourceGeneratorTelemetryCollectorWorkspaceService
+    {
+        private record GeneratorTelemetryKey
+        {
+            public GeneratorTelemetryKey(ISourceGenerator generator)
+            {
+                Identity = new SourceGeneratorIdentity(generator);
+                FileVersion = FileVersionInfo.GetVersionInfo(generator.GetGeneratorType().Assembly.Location).FileVersion ?? "(null)";
+            }
+
+            // TODO: mark these 'required' when we have the attributes in place
+            public SourceGeneratorIdentity Identity { get; init; }
+            public string FileVersion { get; init; }
+        }
+
+        /// <summary>
+        /// Cache of the <see cref="GeneratorTelemetryKey"/> for a generator to avoid repeatedly reading version information from disk;
+        /// this is a ConditionalWeakTable so having telemetry for older runs doesn't keep the generator itself alive.
+        /// </summary>
+        private readonly ConditionalWeakTable<ISourceGenerator, GeneratorTelemetryKey> _generatorTelemetryKeys = new ConditionalWeakTable<ISourceGenerator, GeneratorTelemetryKey>();
+
+        private readonly StatisticLogAggregator<GeneratorTelemetryKey> _elapsedTimeByGenerator = new StatisticLogAggregator<GeneratorTelemetryKey>();
+        private readonly StatisticLogAggregator<GeneratorTelemetryKey> _producedFilesByGenerator = new StatisticLogAggregator<GeneratorTelemetryKey>();
+
+        private GeneratorTelemetryKey GetTelemetryKey(ISourceGenerator generator) => _generatorTelemetryKeys.GetValue(generator, static g => new GeneratorTelemetryKey(g));
+
+        public void CollectRunResult(GeneratorDriverRunResult driverRunResult, GeneratorDriverTimingInfo driverTimingInfo)
+        {
+            foreach (var generatorTime in driverTimingInfo.GeneratorTimes)
+            {
+                _elapsedTimeByGenerator.AddDataPoint(GetTelemetryKey(generatorTime.Generator), generatorTime.ElapsedTime);
+            }
+
+            foreach (var generatorResult in driverRunResult.Results)
+            {
+                _producedFilesByGenerator.AddDataPoint(GetTelemetryKey(generatorResult.Generator), generatorResult.GeneratedSources.Length);
+            }
+        }
+
+        public void ReportStatisticsAndClear(FunctionId functionId)
+        {
+            foreach (var pair in _elapsedTimeByGenerator)
+            {
+                var telemetryKey = pair.Key;
+                var elapsedTimeCounter = pair.Value;
+
+                // We'll log one event per generator
+                Logger.Log(functionId, KeyValueLogMessage.Create(map =>
+                {
+                    map[nameof(telemetryKey.Identity.AssemblyName) + "Hashed"] = AnalyzerNameForTelemetry.ComputeSha256Hash(telemetryKey.Identity.AssemblyName);
```

Comment:
At this point I'm hashing everything which I don't like; in the analyzer case we treat built-in analyzers specially and don't do this, but I don't think I have a way to get an equal control in place. I think we could probably do something like look at the public key to determine if it's a Microsoft produced generator, but not sure if that actually works. While we sort this out, we'll go for this pessimistic approach at the moment.

Reply:
do we need this?  can't we just mark teh bucket a pii information?

Reply:
@CyrusNajmabadi I'm matching what we do for analyzers here; it's entirely possible this can be simplified with that newer support. Happy to investigate further and try to fix both at once.